godot-core/src/engine/gfile.rs

```diff
@@ -0,0 +1,824 @@
+use crate::gen::central::global::Error;
+use crate::gen::classes::FileAccess;
+use crate::obj::{Gd, Inherits};
+
+use std::cmp;
+use std::io::{BufRead, Error as IoError, ErrorKind, Read, Result as IoResult, Seek, Write};
+
+use super::RefCounted;
+
+/// Open file for reading or writing.
+///
+/// It's a convienience wrapper around [`FileAccess`] pointer with unique
+/// reference, providing multiple both safety and quality of life upgrades over
+/// the inner type.
+///  
+/// - Exposes reading and writing capabilities of [`FileAccess`] in a safe way,
+/// returning [`Result<T>`](std::io::Result) instead of the `T` itself.
+/// - Makes the [`FileAccess`] exclusive to its instance, not allowing parallel
+/// reads and writes, which could introduce hard to track bugs by unpredicted
+/// cursor movement. Exclusivity also ensures that when the `GFile` moves out of
+/// scope, inner [`FileAccess`] does also, automatically closing the file.
+/// Alternatively, you can [drop()] the `GFile` to close the file manually.
+/// - Implements useful Rust traits allowing better file processing, namely
+/// [`Read`], [`BufRead`], [`Write`], [`Seek`] and making it better integrated
+/// with various tools in Rust ecosystem (eg. serialization tools).
+///
+/// Files by default are always opened with little-endian, as most files are
+/// saved as such. To switch into big-endian, use `GFile.set_big_endian()`.
+///
+/// ## [`ModeFlags`]
+///
+/// Every constructor opening the connection to file (`open_*` associated
+/// function) intakes the `flags` argument, opening the file for different types
+/// of operations. Regardless of provided `flags` value, cursor is always
+/// positioned at the beginning of the file upon opening. To adjust its position
+/// use [`Seek`]-provided methods.
+///
+/// - `ModeFlags::READ` opens the file for read operations.
+/// - `ModeFlags::WRITE` opens the file for write operations. If the file
+/// doesn't exist at provided `path`, it is created. If it exists, it is
+/// truncated after the file is closed.
+/// - `ModeFlags::READ_WRITE` opens the file for read and write operations.
+/// The file is not truncated after closing.
+/// - `ModeFlags::WRITE_READ` opens the file for read and write operations.
+/// If the file doesn't exist at provided `path`, it is created. If it exists,
+/// it is truncated after the file is closed.
+///
+/// ## Examples
+///
+/// ```no_run
+/// use godot::engine::GFile;
+/// use godot::engine::file_access::ModeFlags;
+/// use godot::builtin::GString;
+///
+/// fn save_game() -> Result<(), std::io::Error> {
+///
+///     // Open file in write mode
+///     let mut my_file = GFile::open("user://save_game.sav", ModeFlags::WRITE)?;
+///
+///     // Write some lines into it
+///     my_file.write_gstring_line("This is my saved game")?;
+///     my_file.write_gstring_line("I played for 5 minutes")?;
+///
+///     Ok(())
+///     // When GFile gets out of scope, the file connection is closed.
```

Comment:
```suggestion
///     // When GFile gets out of scope, the file is closed.
```
"Connection" is more common for something like a database.